gobblin-core/src/main/java/gobblin/writer/http/AbstractAsyncDataWriter.java

```diff
@@ -0,0 +1,179 @@
+package gobblin.writer.http;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import gobblin.writer.AsyncDataWriter;
+import gobblin.writer.WriteCallback;
+import gobblin.writer.WriteResponse;
+
+
+@ThreadSafe
+public abstract class AbstractAsyncDataWriter<D> implements AsyncDataWriter<D> {
+  public static int DEFAULT_BUFFER_CAPACITY = 10000;
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractAsyncDataWriter.class);
+
+  /** Main lock guarding all access */
+  private final ReentrantLock lock;
+
+  /** Condition for waiting takes */
+  private final Condition notEmpty;
+
+  /** Condition for waiting empty */
+  private final Condition isEmpty;
+
+  /** Condition for waiting puts */
+  private final Condition notFull;
+
+  /** Queue to batch records */
+  private final Queue<BufferedRecord<D>> buffer;
+
+  /** Queue capacity */
+  private final int capacity;
+
+  private final ExecutorService service;
+  private Future<?> recordProcessorFuture;
+
+  private volatile boolean isClosing;
+
+  public AbstractAsyncDataWriter(int capacity) {
+    service = Executors.newFixedThreadPool(1);
+    recordProcessorFuture = service.submit(new RecordProcessor());
+
+    lock = new ReentrantLock();
+    notEmpty = lock.newCondition();
+    isEmpty = lock.newCondition();
+    notFull =  lock.newCondition();
+    buffer = new ArrayDeque<>(capacity);
+    this.capacity = capacity;
+    isClosing = false;
+  }
+
+  /**
+   * Dispatch the record in the buffer. Retries should be done if necessary
+   *
+   * @param buffer The buffer which contains a collection of records
+   * @return the number of records are successfully dispatched or -1 if dispatch failed
+   */
+  protected abstract int dispatch(Queue<BufferedRecord<D>> buffer) throws Throwable;
```

Comment:
However the actual dispatch consumes the buffer, whether it consumes one or a batch, is totally up to the subclass implementation. The clarification should be made in subclass not in this abstract class.